spacy/gold/example.pyx

```diff
@@ -26,6 +26,40 @@ cpdef Doc annotations2doc(vocab, tok_annot, doc_annot):
     return output
 
 
+def validate_examples(examples, method):
+    """Check that a batch of examples received during processing is valid.
+    This function lives here to prevent circular imports.
+
+    examples (Iterable[Examples]): A batch of examples.
+    method (str): The method name to show in error messages.
+    """
+    if not isinstance(examples, IterableInstance):
+        err = Errors.E978.format(name=method, types=type(examples))
+        raise TypeError(err)
+    wrong = set([type(eg) for eg in examples if not isinstance(eg, Example)])
+    if wrong:
+        err = Errors.E978.format(name=method, types=wrong)
+        raise TypeError(err)
+
+
+def iter_get_examples(get_examples, name, method="begin_training"):
```

Comment:
Hmm I don't really love this function. I'd rather have the `begin_training` methods doing something like:

```
check_get_examples_is_callable(get_examples, name)
for example in get_examples():
    check_is_example_object(example, name)
```

Reply:
We could call it `validate_get_examples`?

The problem is that if we split this up, we'd have to refactor a bunch of code in different places that assumes it receives an iterable of examples, like `TransitionSystem.get_actions` for instance. Or we'd need _another_ wrapper function for this case that creates a generator of examples, similar to what we have now, so we can check them as they come in.

Reply:
Ah I forgot that we pass the callable around.

Maybe it's better not to make this check in the components? We can do it in the `Language` class which will provide the good user-experience.

Reply:
Not sure, but I thought the idea was to have it in the components as well, because in theory, they could be called spearately? It's also what we currently do for the other methods that take `Example` objects.

(It would be a lot easier without but it'd also mean that we could potentially end up with uncaught errors. So I thougth we wanted it to be clean and consistent.)

Reply:
I think it's okay to be a little less strict in how we check inputs within the components, as if people are calling the components directly we can have a bit more trust that they're passing in the right types.